src/passes/TypeSSA.cpp

```diff
@@ -210,41 +212,28 @@ struct TypeSSA : public Pass {
         existingTypeNames.insert(newName);
       }
     }
-
-    // TODO: arrays
   }
 
-  // An interesting StructNew, which we think is worth creating a new type for,
-  // is one that can be optimized better with a new type. That means it must
-  // have something interesting for optimizations to work with.
+  // An interesting *.new, which we think is worth creating a new type for, is
+  // one that can be optimized better with a new type. That means it must have
+  // something interesting for optimizations to work with.
   //
   // TODO: We may add new optimizations in the future that can benefit from more
   //       things, so it may be interesting to experiment with considering all
   //       news as "interesting" when we add major new type-based optimization
   //       passes.
-  bool isInteresting(StructNew* curr) {
+  bool isInteresting(Expression* curr) {
     if (curr->type == Type::unreachable) {
       // This is dead code anyhow.
       return false;
     }
 
-    if (curr->isWithDefault()) {
-      // This starts with all default values - zeros and nulls - and that might
-      // be useful.
-      //
-      // (A struct whose fields are all bottom types only has a single possible
-      // value in each field anyhow, so that is not interesting, but also
-      // unreasonable to occur in practice as other optimizations should handle
-      // it.)
-      return true;
-    }
-
-    // Look for at least one interesting operand.
-    auto& fields = curr->type.getHeapType().getStruct().fields;
-    for (Index i = 0; i < fields.size(); i++) {
-      assert(i <= curr->operands.size());
-      auto* operand = curr->operands[i];
-      if (operand->type != fields[i].type) {
+    // Look for at least one interesting operand. We will consider each operand
+    // against the declared type, that is, the type declared for where it is
+    // stored. If it has more information than the declared type then it is
+    // interesting.
+    auto isInterestingRelevantTo = [&](Expression* operand, Type declaredType) {
```

Comment:
I'm not sure this is worth pulling out into a lambda rather than just having `operand->type != declaredType` inline.

Reply:
The lambda also contains the `PossibleConstantValues` logic. The type might be the same, but if it is a constant or a global then we also consider it interesting.